Orange/widgets/unsupervised/owsom.py

```diff
@@ -184,6 +187,96 @@ def disconnected_spin(spin):
 N_ITERATIONS = 200
 
 
+class SomSharedValueCompute:
+    def __init__(self, domain: Domain, model: SOM,
+                 offsets: Union[np.ndarray, None],
+                 scales: Union[np.ndarray, None]):
+        # offsets and scales are made immutable so that they are hashable
+        def immutable(a):
+            if a is not None:
+                a = a.copy()
+                a.flags.writeable = False
+            return a
+
+        self.domain = domain
+        self.model = model
+        self.offsets = immutable(offsets)
+        self.scales = immutable(scales)
+        self.__hash = None
+
+    def __call__(self, data):
+        x = data.transform(self.domain).X
+        cont_x, mask, _, _ = SOM.prepare_data(x, self.offsets, self.scales)
+        winners = np.full((len(data), 2), np.nan)
+        distances = np.full(len(data), np.nan)
+        winners[mask], distances[mask] = self.model.winners(cont_x)
+        winners += 1
+        return winners, distances
+
+    # `offsets` and `scales` are ndarray's (and `model` contains ndarray's)
+    # Unless we __eq__ by identity, we can't properly define hash.
+    def __eq__(self, other):
+        return type(self) is type(other) and \
+               self.domain == other.domain \
+               and self.model is other.model \
+               and np.array_equal(self.offsets, other.offsets) \
+               and np.array_equal(self.scales, other.scales)
+
+    def __hash__(self):
+        if self.__hash is None:
+            self.__hash = hash(
+                (hash(self.domain),
+                 id(self.model),
+                 hash(tuple(self.offsets)), hash(tuple(self.scales))))
```

Comment:
`__hash__` can be called often so unless these arrays are always small, this line can be slow.

Reply:
I'd just use some elements from the arrays for speed and not worry about hash conflicts, which can happen anyway. Can't hurt much if they come up a bit more often.

Reply:
The size of the array equals the number of numeric attributes. SOM itself has a time complexity of #variables * #instances * #iterations, hence the computation of hash should be negligible in comparison of computing SOM. Also, the value is cached and hence computed only once per a SOM model - and we won't have many of them.

Fun fact: you can't use the `@cache` decorator on `__hash__`. :)

```
>>> from functools import cache
>>> class A:
...     @cache
...     def __hash__(self):
...         return 42
...
>>> a = A()
>>> hash(a)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
RecursionError: maximum recursion depth exceeded while calling a Python object
```